### PR TITLE
Avoid using the stack for large memory block allocation

### PR DESCRIPTION
**Pull request by benjamreis**
This can lead to memory corruption, use calloc/free instead.

Original commits from Andrei Semenov <andrei.semenov@vates.fr>:
- https://github.com/xcp-ng/uefistored/pull/41/commits/f4d54998fd3ec866cd2c597e972df16fce47805d
- https://github.com/xcp-ng/uefistored/pull/41/commits/e4a936cd5a734980214223bbc3e9b6b8de9e1ea9

Co-author; Andrei Semenov <andrei.semenov@vates.fr>
Signed-off-by: BenjiReis <benjamin.reis@vates.fr>

### CODE REVIEW

**Comment by benjamreis** [comment]
Unit tests successfull :+1:

File: src/xapi.c [hunk 1]
```diff
@@ -632,19 +632,29 @@ static void throttle(void)
  */
 int xapi_set(void)
 {
-    char buffer[MSG_SIZE];
+    char* buffer;
```
Comment (Wescoeur) [inline src/xapi.c:635]: Coding style: `char *buffer` instead. :wink: (There are other declaration in the same case.)

File: src/xapi.c [hunk 2]
```diff
@@ -632,19 +632,29 @@ static void throttle(void)
...
     int ret;
 
     throttle();
 
+    buffer = calloc(1, MSG_SIZE);
```
Comment (Wescoeur) [inline src/xapi.c:640]: You can declare the variable directly here.

File: src/xapi.c [hunk 3]
```diff
@@ -692,30 +702,43 @@ int xapi_connect(void)
...
     int hdr_len;
     size_t body_len;
     int ret;
 
+    message = (char*)calloc(2, MSG_SIZE);
+    if (!message) {
+        return -ENOMEM;
+    }
+
+    body = message + MSG_SIZE;
```
Comment (Wescoeur) [inline src/xapi.c:716]: You can declare the body variable here. (Same for the message variable.)

File: src/xapi.c [hunk 4]
```diff
@@ -692,30 +702,43 @@ int xapi_connect(void)
...
+    }
 
     body_len = ret;
 
     hdr_len = create_header(body_len, message, MSG_SIZE);
 
     if (hdr_len < 0) {
-        return -1;
+        ret = -1;
+	goto out;
```
Comment (Wescoeur) [inline src/xapi.c:732]: Bad indentation.

File: src/xapi.c [hunk 5]
```diff
@@ -1094,7 +1117,12 @@ int base64_from_response(char *buffer, size_t n, char *response)
 static int xapi_get_nvram(char *session_id, char *buffer, size_t n)
 {
     int status;
-    char response[MSG_SIZE] = { 0 };
+    char* response;
+
+    response = (char*)calloc(1, MSG_SIZE);
```
Comment (Wescoeur) [inline src/xapi.c:1122]: Declare response directly here. :p

File: src/xapi.c [hunk 6]
```diff
@@ -1109,16 +1137,19 @@ static int xapi_get_nvram(char *session_id, char *buffer, size_t n)
 
     if (status != 0) {
         ERROR("VM.get_NVRAM failed: status=%d\n", status);
-        return -1;
+        status =  -1;
```
Comment (Wescoeur) [inline src/xapi.c:1140]: `status = -1`

File: src/xapi.c [hunk 7]
```diff
@@ -1148,21 +1178,34 @@ int xapi_variables_request(variable_t *vars, size_t n)
         return 0;
     }
 
+    plaintext = (uint8_t*)calloc(2, MSG_SIZE);
+    if (!plaintext) {
+        ERROR("failed to allocate memory\n");
+        return -ENOMEM;
+    }
+
+    b64 = (char*)(plaintext + MSG_SIZE);
```
Comment (Wescoeur) [inline src/xapi.c:1187]: b64/plaintext declaration. :wink:

File: src/xapi.c [hunk 8]
```diff
@@ -632,19 +632,29 @@ static void throttle(void)
...
 int xapi_set(void)
 {
-    char buffer[MSG_SIZE];
+    char* buffer;
     int ret;
 
     throttle();
 
+    buffer = calloc(1, MSG_SIZE);
+    if (!buffer)  {
```
Comment (Wescoeur) [inline src/xapi.c:641]: `if (!buffer) {`

File: src/xapi.c [hunk 9]
```diff
@@ -692,30 +702,43 @@ int xapi_connect(void)
...
     va_list ap;
-    char message[MSG_SIZE];
-    char body[MSG_SIZE];
+    char* message;
+    char* body;
     int hdr_len;
     size_t body_len;
     int ret;
 
+    message = (char*)calloc(2, MSG_SIZE);
```
Comment (Wescoeur) [inline src/xapi.c:711]: `(char *)calloc(2, MSG_SIZE);`

File: src/xapi.c (hunk 5 above)
Comment (Wescoeur) [reply src/xapi.c:1122]: And: `response = (char *)calloc(1, MSG_SIZE);`

File: src/xapi.c (hunk 7 above)
Comment (Wescoeur) [reply src/xapi.c:1187]: And `(char *)(plaintext + MSG_SIZE);`

**Comment by stormi** [comment]
I'm good with the principle of the changes. I'm still hesitating regarding the merge into the next release of uefistored that will be included in XCP-ng 8.2.1, though, as I'd like to avoid any last minute regression and we currently can't reproduce any real-world user bug related to the on the stack allocation. On the other hand, it's in-depth security so good to have.

In any case, everyone please put your best analysis and coding skills on this in order to reduce the regression risks to the maximum. This will make the decision simpler.

**Comment by benjamreis** [comment]
XCP-ng tests successfull on Alpine, Debian 10 & Win10 (with `efitools`)

File: src/xapi.c [hunk 10]
```diff
@@ -1148,21 +1173,34 @@ int xapi_variables_request(variable_t *vars, size_t n)
         return 0;
     }
 
+    plaintext = (uint8_t*)calloc(2, MSG_SIZE);
```
Comment (Wescoeur) [inline src/xapi.c:1176]: `uint8_t *plaintext = (uint8_t *)calloc(2, MSG_SIZE);` :sweat_smile:

**Review by Wescoeur — APPROVED** [review]

**Review by stormi — APPROVED** [review]

**Review by andSmv — APPROVED** [review]

**Pull request merged**